File: highway-pursuit-server/highway-pursuit-server/Injected/ScoreService.cs
using HighwayPursuitServer.Server;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace HighwayPursuitServer.Injected
{
    class ScoreService
    {
        private readonly IHookManager _hookManager;
        private int _lastScore = 0;
        private int _newScore = 0;

        public ScoreService(IHookManager hookManager)
        {
            this._hookManager = hookManager;
            this.RegisterHooks();
        }

        public int PullReward()
        {
            var scoreDelta = _newScore - _lastScore;
            _lastScore = _newScore;
            return scoreDelta;
        }

        public void ResetScore()
        {
            _lastScore = 0;
            _newScore = 0;
        }

        #region Hooking
        private void RegisterHooks()
        {
            IntPtr setScorePtr = new IntPtr(_hookManager.GetModuleBase().ToInt32() + MemoryAdresses.SET_SCORE_OFFSET);
            SetScore = Marshal.GetDelegateForFunctionPointer<SetScore_delegate>(setScorePtr);
            _hookManager.RegisterHook(setScorePtr, new SetScore_delegate(SetScore_Hook));
        }

        #region hooks
        void SetScore_Hook(int score)
        {
            SetScore(score);
            _newScore = score;
        }
        #endregion

        #region delegates
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        delegate void SetScore_delegate(int score);
        #endregion

        #region original function pointers
        static SetScore_delegate SetScore;
        #endregion
        #endregion
    }
}

File: highway-pursuit-server-2/highway-pursuit-server/Injected/ScoreService.cpp
#include "../pch.h"
#include "ScoreService.hpp"
#include "MemoryAddresses.hpp"

namespace Injected
{
    ScoreService::ScoreService(std::shared_ptr<HookManager> hookManager)
        : _hookManager(hookManager)
    {
        RegisterHooks();
    }

    uint32_t ScoreService::PullReward()
    {
        uint32_t scoreDelta = _newScore - _lastScore;
        _lastScore = _newScore;
        return scoreDelta;
    }

    // RegisterHooks method
    void ScoreService::RegisterHooks()
    {
        ScoreService::Instance = this;

        LPVOID setScorePtr = reinterpret_cast<LPVOID>(_hookManager->GetModuleBase() + MemoryAddresses::SET_SCORE_OFFSET);
        _hookManager->RegisterHook(setScorePtr, &SetScore_StaticHook, &SetScore_Base);
    }

    void ScoreService::SetScore_Hook(uint32_t score)
    {
        if (score == 0)
        {
            // If SetScore is called with 0, a reset happended
            _lastScore = 0;
        }
        _newScore = score;
        SetScore_Base(score);
    }

    void __cdecl ScoreService::SetScore_StaticHook(uint32_t score)
    {
        if (ScoreService::Instance != nullptr)
        {
            return ScoreService::Instance->SetScore_Hook(score);
        }
    }

    ScoreService* ScoreService::Instance = nullptr;
    ScoreService::SetScore_t ScoreService::SetScore_Base = nullptr;
}
